### ISSUE TITLE
False unreachable warning on match statement

### ISSUE THREAD

**Issue by delfick** [python/mypy#14485]
**Bug Report**

Hello,

I want to do something like:

```python
def match(info: RequestContext):
    match info:
        case RequestContext.ForPopup(data_only_in_popup_context=data):
            print(data)
        case RequestContext.ForDetailPage(data_only_in_detail_page=data):
            print(data)
```

However I'm getting unreachable warnings when --warn-unreachable is turned on. But the code I have does indeed reach that line.

**To Reproduce**

https://gist.github.com/delfick/e7974027a422f8f212887e490e5d3a75

**Expected Behavior**

I expect no errors when running with mypy and the warn-unreachable option.

**Actual Behavior**

When I run `python example.py` I get expected output

```
For popup <__main__.Request object at 0x10326a620> (1,) {'1': '2'}
For details page <__main__.Request object at 0x10326a0b0> {'3': '4'}
For popup {'1': '2'}
For details page {'3': '4'}
```

When I run `mypy example.py --warn-unreachable` I get

```
example.py:49: error: Statement is unreachable  [unreachable]
example.py:65: error: Statement is unreachable  [unreachable]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.991
- Mypy command-line flags: --warn-unreachable
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.10.8

**Comment by delfick** [comment]
for the benefit of anyone who comes across this, I solved my original problem with a slightly different approach, https://gist.github.com/delfick/dddd2e59638cd807584d0d6fe99a7d13. But I still think it would be nice if my other approach didn't fail the way it does :)

**Comment by thierry-martinez** [comment]
I have the following example that is shorter and exhibits a very similar issue (I don't know if I should fill another issue, because the cause is probably the same; if you think I should open another issue, feel free to ask!):
```
import typing


class C(typing.NamedTuple):
    v: str


def f(x: typing.Any) -> None:
    match x:
        case C(v=v):
            print(v)
        case [y]:
            print("HERE: " + y)


f(["hello"])
```
The code prints `HERE: hello` as expected, so the line `print("HERE: " + y)` is definitively reachable, but `mypy --warn-unreachable` says the contrary.

**Comment by A5rocks** [comment]
Later smaller case is fixed now. First might be similar to https://github.com/python/mypy/issues/19981? I haven't looked too much at the reproducer though.

**Comment by A5rocks** [comment]
I think this might be intentional. If you look at the attributes for `RequestContext`, mypy doesn't have a concept that `url_args` (for instance) is potentially undefined.

I think that the best current representation is that `url_args` is actually `tp.Sequence[object] | Never`. Then we still mark the second branch as unreachable! I think this is more an issue with `runtime_checkable` than with the protocols at hand...

I'm very unsure, so I'll just mark this as pending rather than closing immediately.

**Comment by hauntsaninja** [comment]
Yeah, A5rocks is correct

thierry-martinez's case was fixed by #18448
